keps/20190218-add-testing-infrastructure.md

```diff
@@ -0,0 +1,141 @@
+---
+kep-number: 4
+title: Add Testing Infrastructure
+authors:
+  - "@runyontr"
+owners:
+  - @runyontr
+  - "@fabian"
+editor: TBD
+creation-date: 2019-02-18
+last-updated: 2010-02-18
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# add-testing-infrastructure
+
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories](#user-stories)
+      * [Story 1](#story-1)
+      * [Story 2](#story-2)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+As the complexity and scope of KUDO grows, it becomes impossible to manually validate 
+exisiting frameworks and capabilities still function as expected.  As part of a robust 
+CICD pipeline, a testing suite needs to be developed that can validate each commit, pull 
+request, or even deployment of KUDO on a cluster.
+
+These testing packages should also assure conformance for the `kudobuilder/frameworks` repository 
+to particular versions of KUDO.
+
+
+
+## Motivation
+
+### Goals
```

Comment:
Additionally, do we want to create a definition of "mergeable" for other stories to set the baseline amount of testing required to merge future features once we pull ourselves out of the hole we've given ourselves? Is "tests must be included for new functionality" enough? Is it "overall code coverage can not go down in new PRs without entire group approval (for hard to test things)"?

Reply:
I think it would be good to have a definition/requirement like this.  The downside of "code coverage" is now that's something we need to be able to track and compare between commits.

Depending on our CI tool, that could in theory be a warning that says "the code coverage went down" to give reviewers insight.